sub-packages/bionemo-dnadl/src/bionemo/dnadl/io/dataset.py

```diff
@@ -0,0 +1,189 @@
+# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
+#
+# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
+# property and proprietary rights in and to this material, related
+# documentation and any modifications thereto. Any use, reproduction,
+# disclosure or distribution of this material and related documentation
+# without an express license agreement from NVIDIA CORPORATION or
+# its affiliates is strictly prohibited.
+
+import os
+from abc import ABC
+from pathlib import Path
+from typing import Any, Callable, List, Optional
+
+import polars as pl
+import pysam
+import torch
+from torch.utils.data import Dataset
+
+from bionemo.dnadl.tools.genome_editing import create_personal_sequence
+from bionemo.dnadl.tools.genome_interval import GenomeInterval
+from bionemo.dnadl.tools.vcf import SampleId, read_variants_in_interval
+from bionemo.noodles.nvfaidx import NvFaidx
+
+
+DNATokenizer = Callable[[str], torch.Tensor]
+
+
+class Genome:
+    """Class that creates a genome object from a fasta file. It can be used to extract sequences from the genome."""
+
+    def __init__(self, fasta_file: str | os.PathLike):
+        """Instantiate the class."""
+        fasta_file = Path(fasta_file)
+        assert fasta_file.exists(), "path to fasta file must exist"
+        self.seqs = NvFaidx(str(fasta_file))
+
+    def adjust_interval(self, interval: GenomeInterval, context_length: int) -> GenomeInterval:
+        """Used to extend an interval by a fixed amount on both sides."""
+        start, end = interval.start, interval.end
+        interval_length = interval.end - interval.start
+        chromosome_length = self.seqs[interval.chromosome].length
+        if interval_length >= context_length:
+            return interval
+
+        extra_seq = context_length - interval_length
+        extra_left_seq = extra_seq // 2
+        extra_right_seq = extra_seq - extra_left_seq
+
+        start -= extra_left_seq
+        end += extra_right_seq
+
+        if start <= 0:
+            start = 1
+
+        if end > chromosome_length:
+            end = chromosome_length
+
+        return GenomeInterval(interval.chromosome, start, end)
+
+    def extract_sequence(self, genome_interval: GenomeInterval) -> str:
+        """Extract a sequence from the genome."""
+        return str(self.seqs[genome_interval.chromosome][genome_interval.start : genome_interval.end])
+
+
+class GenomeDataset(Dataset, ABC):
+    """Abstract class for datasets that return a DNA sequence."""
+
+    def __init__(self, tokenizer: DNATokenizer):
+        """Instantiate the class."""
+        self.tokenizer = tokenizer
+
+    def __getitem__(self, ind: int) -> torch.tensor:
+        """Get an item from the dataset."""
+        sequence = self.get_dna_sequence(ind)
+        return self.tokenizer(sequence)
+
+    def get_dna_sequence(self, ind: int) -> str:
+        """Get the DNA sequence."""
+        ...
```

Comment:
Could you expand the docstrings here? no specific feedback other than adding arguments, but here's my perspective as a reader: 

- the constructor implies this is doing a transformation with the tokenizer, like a `MappedDataset`
- it looks like `ind` is meant to provide a deterministic mapping from dataset 'index' to some DNA sequence.
- I can think of how I might want to use this, but maybe providing a few concise examples in text would help future implementors (or just point them at your implementations below).